### PR TITLE
feat: QuickBooks (+ refactoring `postman2openapi` script into a provider)

### PR DESCRIPTION
**Pull request by RafalWilinski**
Closes https://github.com/Stedi/mappings/issues/244

Introduces generic `PostmanProvider` which fetches Postman collections from [Stedi's registry Postman Workspace](https://www.postman.com/stedi-inc/workspace/registry/), transforms to an OpenAPI and extracts JSONSchemas from response and request objects.

### CODE REVIEW

File: src/providers/postman.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    return this.versions;
+  }
+
+  /**
+   * Fetches Postman collection based on the postmanCollectionId and transforms it to OpenAPI definition
+   * @returns OpenAPI3Schema object containing the OpenAPI definition with the entities to be processed
+   */
+  async getSchema(version: string): Promise<OpenAPI3Schema> {
+    if (!this.postmanCollectionId) {
+      throw new Error('Neither "postmanCollectionId" wasn\'t provided, nor method "getSchema" was overriden');
```
Comment (bahrmichael) [inline src/providers/postman.ts:91]: Typos: Neither "postmanCollectionId" was provided [double negotiation], nor method "getSchema" was overridden.

File: src/providers/postman.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    this.docsLink = docsLink;
+    this.customPath = customPath;
+    this.postmanCollectionId = postmanCollectionId;
+  }
+
+  /**
+   * Returns status of provider. Enabled only if POSTMAN_API_KEY environement variable is set.
+   */
+  isEnabled(): boolean {
+    return !!this.postmanCollectionId;
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:74]: Nit: I think `this.postmanCollectionId != null` would be more readable and accessible for people that are not aware of how the type coercion works in Javascript.

File: src/providers/postman.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  async getVersions(): Promise<string[]> {
+    return this.versions;
+  }
+
+  /**
+   * Fetches Postman collection based on the postmanCollectionId and transforms it to OpenAPI definition
+   * @returns OpenAPI3Schema object containing the OpenAPI definition with the entities to be processed
+   */
+  async getSchema(version: string): Promise<OpenAPI3Schema> {
+    if (!this.postmanCollectionId) {
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:90]: Should not we use the `isEnabled` function here?

File: src/providers/postman.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    return {
+      type: "openapi-v3",
+      versionName: version,
+      value: jsonOpenApiSchema,
+      entities: this.entities,
+    };
+  }
+
+  async unbundle({ value }: OpenAPI3Schema): Promise<EntitySchema[]> {
+    const results: JSONSchema7[] = [...extractRequestBodies(value), ...extractResponseSchemas(value)];
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:113]: Are not we using jsonschema2020?

File: src/providers/postman.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+    return results.map((value) => ({
+      name: value.title as string,
+      schema: this.sanitizeSchemaFunction ? this.sanitizeSchemaFunction(value) : value,
+    }));
+  }
+}
+
+/**
+ * Goes through POST requests and extracts request bodies as JSONSchemas
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:123]: Nit: In my humble opinion this comment is not necessary as it rephrases the function name. It does not add any context to the WHY of the function.

File: src/providers/postman.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+      results.push(entitySchema);
+    }
+  }
+
+  return results;
+}
+
+/**
+ * Goes through GET requests and extracts response bodies as JSONSchemas
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:150]: As per my previous comment.

File: src/providers/postman.ts [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    this.versions = versions;
+    this.entities = entities;
+    this.sanitizeSchemaFunction = sanitizeSchema;
+    this.docsLink = docsLink;
+    this.customPath = customPath;
+    this.postmanCollectionId = postmanCollectionId;
+  }
+
+  /**
+   * Returns status of provider. Enabled only if POSTMAN_API_KEY environement variable is set.
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:71]: `environement` -> `environment`

File: src/providers/postman.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+/**
+ * Goes through POST requests and extracts request bodies as JSONSchemas
+ */
+function extractRequestBodies(value: unknown): JSONSchema7[] {
+  const results: JSONSchema7[] = [];
+
+  for (const [_path, info] of Object.entries((value as any).paths)) {
+    if ((info as any).post) {
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:129]: Nit: I think the code would read better if we would not indent the main bulk of the logic.
We can escape early here if the `info.post` does not exist.

```
const hasPost = "post" in (info as any);
    if (!hasPost) {
      continue;
    }
    // rest of the code
```

File: src/providers/postman.ts [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+/**
+ * Goes through GET requests and extracts response bodies as JSONSchemas
+ */
+function extractResponseSchemas(value: unknown): JSONSchema7[] {
+  const results: JSONSchema7[] = [];
+
+  for (const [_path, info] of Object.entries((value as any).paths)) {
+    if ((info as any).get) {
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:156]: As per my previous comment

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -42,15 +42,55 @@ export class MyCompanyNewProvider extends OpenAPIProvider {
 }
 ```
 
+After writing your provider, please also add it to the `src/providers/index.ts` file.
+
+#### Postman-based Provider
+
+If the API you want to add to the Registry exposes [Postman](https://www.openapis.org) specification, you can add it to the registry in two steps:
```
[the remark this answers is not in the document]
Comment (RafalWilinski) [reply CONTRIBUTING.md:49]: Yes! My bad

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -42,15 +42,55 @@ export class MyCompanyNewProvider extends OpenAPIProvider {
...
+
+1. Import desired collection to the Stedi's ["registry" workspace](https://www.postman.com/stedi-inc/workspace/registry/overview) in Postman.
```
[the remark this answers is not in the document]
Comment (RafalWilinski) [reply CONTRIBUTING.md:51]: I'd skip a non-Stedi scenario for now - we don't have many contributions from the in-house so true OSS are even less likely for now. I'd defer it for later.

File: src/providers/openapi.ts [hunk 12]
```diff
@@ -37,7 +38,7 @@ export class OpenAPIProvider implements BaseProvider {
   /**
    * Custom function to sanitize the schema, e.g. remove unsupported keywords or custom formats.
    */
-  private readonly sanitizeSchemaFunction?: (schema: unknown) => unknown;
+  public readonly sanitizeSchemaFunction?: (schema: JSONSchema) => JSONSchema;
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/openapi.ts:41]: Nit: Consider renaming to `sanitizeSchema`. I think the suffix `Function` is not necessary here.

Comment (WojciechMatuszewskiStedi) [inline src/providers/openapi.ts:41]: Nit: Consider renaming this method to `sanitizeSchema`. I think the `Function` suffix is not needed.
I would also consider renaming the `sanitizeSchema` standalone function declared in this file to `defaultSanitizeSchema` to indicate that the function is used as a fallback or "default".

File: src/providers/postman.ts [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    this.docsLink = docsLink;
+    this.customPath = customPath;
+    this.postmanCollectionId = postmanCollectionId;
+  }
+
+  /**
+   * Returns status of provider. Enabled only if POSTMAN_API_KEY environment variable is set.
+   */
+  isEnabled(): boolean {
+    return !!this.postmanCollectionId && !!this.postmanApiKey;
```
Comment (WojciechMatuszewskiStedi) [inline src/providers/postman.ts:75]: Nit: consider moving away from `!!` as it uses several JS concepts that might not be obvious to others. I think we should favor explicit equals checks `this.postmanCollectionId != null` here.

**Review by WojciechMatuszewskiStedi — APPROVED** [review]

**Pull request merged**